File: 2972/main.go
package main

import "fmt"

func incremovableSubarrayCount(nums []int) int64 {
	n := len(nums)
	var ans int64 = 0
	l := 0
	for l < n-1 {
		if nums[l] >= nums[l+1] {
			break
		}
		l++
	}
	if l == n-1 {
		return int64(n) * int64(n+1) / 2
	}
	ans += int64(l+2)
	for r := n-1; r > 0; r-- {
		if r < n-1 && nums[r] >= nums[r+1] {
			break
		}
		for l >= 0 && nums[l] >= nums[r] {
			l--
		}
		ans += int64(l+2)
	}
	return ans
}

func main() {
	fmt.Println(incremovableSubarrayCount([]int{1, 2, 3, 4}))
}

File: 2972/main.rs
struct Solution{}

impl Solution {
    pub fn incremovable_subarray_count(nums: Vec<i32>) -> i64 {
        let n = nums.len();
        let mut ans: i64 = 0;
        let mut l: i32 = 0;

        while (l as usize) < n - 1 {
            if nums[l as usize] >= nums[l as usize+1] {
                break;
            }
            l += 1;
        }
        if l as usize == n-1 {
            return (n * (n + 1) / 2) as i64;
        }
        ans += (l+2) as i64;
        let mut r = n-1;
        while r > 0 {
            if r < n-1 && nums[r] >= nums[r+1] {
                break;
            }
            
            while l as i32 >= 0 && nums[l as usize] >= nums[r] {
                l -= 1;
            }
            ans += (l+2) as i64;
            r -= 1;
        }
        ans
    }
}

fn main() {
    println!("{}", Solution::incremovable_subarray_count(vec![1,2,3,4]));
    println!("{}", Solution::incremovable_subarray_count(vec![8,7,6,6]));

}